src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AccessController.cs

```diff
@@ -48,12 +54,25 @@ public class AccessController : Controller
             // which is required for this stateless OAuth2/OIDC token endpoint to work correctly.
             // To prevent effective CSRF/session fixation attacks, this action MUST NOT
             // return an authentication cookie or try to establish an ASP.NET Core session.
+            var openIdSettings = await _openIdService.GetOpenIdSettingsAsync();
+            var application = await _applicationManager.FindByClientIdAsync(request.ClientId);
+            if (application == null)
+            {
+                return View("Error", new ErrorViewModel
+                {
+                    Error = OpenIdConnectConstants.Errors.InvalidClient,
+                    ErrorDescription = T["Details concerning the calling client application cannot be found in the database"]
+                });
+            }
+
+            var isPasswordGrantType = request.IsPasswordGrantType() && openIdSettings.AllowPasswordFlow && application != null && application.AllowPasswordFlow;
+            var isClientCredentialsType = (request.IsClientCredentialsGrantType() && openIdSettings.AllowClientCredentialsFlow && application != null && application.AllowClientCredentialsFlow);
 
-            if (request.IsPasswordGrantType())
+            if (isPasswordGrantType)
             {
                 return await ExchangePasswordGrantType(request);
             }
-            else if (request.IsClientCredentialsGrantType())
+            else if (isClientCredentialsType)
             {
                 return await ExchangeClientCredentialsGrantType(request);
             }
```

Comment:
I though I would need to control here also AuthorizationCodeFlow, HybridFlow and RefreshTokenGrantType cause they use the token endpoint. However execution never pass thought this action. So, at this point when general settings allow those flows but an app doesn't allow them I'm not able to forbid its use here. For AuthorizationCodeFlow and HybridFlow a minimum solution is I can control it in Authorize action, but for RefreshTokenGrantType I'm not able I'm not able to forbid its use if an openid app doesn't allow it.